test/index.ts

```diff
@@ -1006,6 +1007,14 @@ schema = Joi.link(str);
             .then(val => { throw new Error('one error'); })
             .catch(e => { });
 
+        expect.type<Promise<TResult>>(schema.validateAsync(value));
+        expect.type<Promise<{ value: TResult, artifacts: Map<any, string[][]> }>>(schema.validateAsync(value, { artifacts: true }));
+        expect.type<Promise<{ value: TResult, warning: Joi.ValidationError[] }>>(schema.validateAsync(value, { warnings: true }));
+        expect.type<Promise<{ value: TResult, artifacts: Map<any, string[][]>; warning: Joi.ValidationError[] }>>(schema.validateAsync(value, { artifacts: true, warnings: true }));
+        expect.error<Promise<{ value: TResult, warning: Joi.ValidationError[] }>>(schema.validateAsync(value, { artifacts: true }));
+        expect.error<Promise<{ value: TResult, artifacts: Map<any, string[][]> }>>(schema.validateAsync(value, { warnings: true }));
+        expect.error<Promise<TResult>>(schema.validateAsync(value, { artifacts: true, warnings: true }));
```

Comment:
Wouldn't it be interesting also to test when passing `{ artifacts: false }` and `{ warning: false }` as well? To make sure neither `artifacts` or `warning` is included in the result.